solution/server/src/user-data-access.ts

```diff
@@ -0,0 +1,309 @@
+import { User, UserInput, UserListParams, UserPaginatedResponse } from './types/types'
+import { ApolloError } from 'apollo-server-errors'
+import {
+	AttributeValue,
+	DeleteItemInput,
+	GetItemInput,
+	PutItemInput,
+	PutItemInputAttributeMap,
+	ScanInput,
+	DocumentClient,
+	ScanOutput,
+	ItemList,
+	AttributeMap,
+} from 'aws-sdk/clients/dynamodb'
+import moment from 'moment'
+import { v4 as uuidv4 } from 'uuid'
+
+const tableName = process.env.USERS_TABLE_NAME ?? ''
+
+const defaultListLimit = 10
+
+/**
+ * Retrives an User by its ID
+ * @param db The DynamoDB DocumentClient param
+ * @param id The User ID
+ * @returns The User
+ */
+export const GetUserByIdAsync = async (db: DocumentClient, id: string): Promise<User> => {
+	try {
+		const params: GetItemInput = {
+			TableName: tableName,
+			Key: {
+				id: id as AttributeValue,
+			},
+		}
+
+		const userDbObj = await db
+			.get(params)
+			.promise()
+			.catch((err) => {
+				throw err
+			})
+
+		if (!userDbObj || !userDbObj.Item) {
+			const errorMessage = `No user was found`
+			console.log(errorMessage)
+			throw new ApolloError(errorMessage)
+		}
+
+		return attributeMapToUser(userDbObj.Item)
+	} catch (error) {
+		const errorMessage = 'Error on User get'
+		console.log(errorMessage)
+		throw new ApolloError(errorMessage, error)
+	}
+}
+
+/**
+ * Retrieves an user list from the given params, alowing pagination and filtering
+ * @param db The DynamoDB DocumentClient param
+ * @param listingParams The pagination params
+ * @returns The paginated user list
+ */
+export const ListUsersAsync = async (db: DocumentClient, listingParams: UserListParams | null | undefined): Promise<UserPaginatedResponse> => {
+	try {
+		// Set the base limit
+		const limit = listingParams?.limit ?? defaultListLimit
+		const params: ScanInput = {
+			TableName: tableName,
+			Limit: limit,
+		}
+
+		// Check and add if there is a filter
+		if (listingParams?.filter) {
+			params.FilterExpression = 'contains(#name, :filter)'
+			params.ExpressionAttributeNames = {
+				'#name': 'name',
+			}
+			params.ExpressionAttributeValues = {
+				':filter': listingParams.filter as AttributeValue,
+			}
+		}
+
+		// Check if there is a starting point on query, "lastEvaluatedId" is the id of the last visited record
+		if (listingParams?.lastEvaluatedKey) {
+			params.ExclusiveStartKey = {
+				id: listingParams?.lastEvaluatedKey as AttributeValue,
+			}
+		}
+
+		// DynamoDB does not directly support pagination. The "limit" param just limits how many entities will be visited and not how many will be retrieved
+		// That's why we must keep looking until there is no more entities to be visited or the user limit was satisfied
+		// https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Query.Pagination.html
+		let userList: User[] = []
+		let userDbObj: ScanOutput
+
+		do {
+			userDbObj = await db
+				.scan(params)
+				.promise()
+				.catch((err) => {
+					throw err
+				})
```

Comment:
Why do we need to add this catch?

Reply:
Indeed, we really don't need this one, and none of the catches on DocumentClient method calls, because they will already be caught for being inside the try-catch scope. They will not add any errors but should be removed since we are not giving any special treatment in this version for DynamoDB exceptions.